packages/Operators/src/PointCloud/DTK_NodeToNodeOperator_impl.hpp

```diff
@@ -0,0 +1,252 @@
+#include "DTK_SplineInterpolationPairing.hpp"
+#include "DTK_BasicEntityPredicates.hpp"
+#include "DTK_PredicateComposition.hpp"
+#include "DTK_EuclideanDistance.hpp"
+
+#include <Teuchos_CommHelpers.hpp>
+#include <Teuchos_Ptr.hpp>
+#include <Teuchos_ArrayRCP.hpp>
+#include <Teuchos_ParameterList.hpp>
+
+#include <Tpetra_MultiVector.hpp>
+
+namespace DataTransferKit
+{
+//---------------------------------------------------------------------------//
+// Constructor.
+template<int DIM>
+NodeToNodeOperator<DIM>::NodeToNodeOperator(
+    const Teuchos::RCP<const TpetraMap>& domain_map,
+    const Teuchos::RCP<const TpetraMap>& range_map,
+    const Teuchos::ParameterList& parameters )        
+    : Base( domain_map, range_map )
+{ /* ... */ }
+
+//---------------------------------------------------------------------------//
+// Setup the map operator.
+template<int DIM>
+void NodeToNodeOperator<DIM>::setupImpl(
+    const Teuchos::RCP<FunctionSpace>& domain_space,
+    const Teuchos::RCP<FunctionSpace>& range_space )
+{
+    DTK_REQUIRE( Teuchos::nonnull(domain_space) );
+    DTK_REQUIRE( Teuchos::nonnull(range_space) );
+
+    // Extract the Support maps.
+    const Teuchos::RCP<const typename Base::TpetraMap> domain_map
+	= this->getDomainMap();
+    const Teuchos::RCP<const typename Base::TpetraMap> range_map
+	= this->getRangeMap();
+
+    // Get the parallel communicator.
+    Teuchos::RCP<const Teuchos::Comm<int> > comm = domain_map->getComm();
+
+    // Determine if we have range and domain data on this process.
+    bool nonnull_domain = Teuchos::nonnull( domain_space->entitySet() );
+    bool nonnull_range = Teuchos::nonnull( range_space->entitySet() );
+
+    // Extract the source nodes and their ids.
+    EntityIterator domain_iterator;
+    if ( nonnull_domain )
+    {
+	LocalEntityPredicate local_predicate(
+	    domain_space->entitySet()->communicator()->getRank() );	
+	PredicateFunction domain_predicate =
+	    PredicateComposition::And(
+		domain_space->selectFunction(),	local_predicate.getFunction() );
+	domain_iterator =
+	    domain_space->entitySet()->entityIterator( 0, domain_predicate );
+    }
+    int local_num_src = domain_iterator.size();
+    Teuchos::ArrayRCP<double> source_centers( DIM*local_num_src);
+    Teuchos::ArrayRCP<GO> source_support_ids( local_num_src );
+    Teuchos::Array<SupportId> source_node_supports;
+    EntityIterator domain_begin = domain_iterator.begin();
+    EntityIterator domain_end = domain_iterator.end();
+    int entity_counter = 0;
+    for ( EntityIterator domain_entity = domain_begin;
+	  domain_entity != domain_end;
+	  ++domain_entity, ++entity_counter )
+    {
+	domain_space->shapeFunction()->entitySupportIds(
+	    *domain_entity, source_node_supports );
+	DTK_CHECK( 1 == source_node_supports.size() );
+	source_support_ids[entity_counter] = source_node_supports[0];
+	domain_space->localMap()->centroid(
+	    *domain_entity, source_centers(DIM*entity_counter,DIM) );
+    }
+
```

Comment:
you are right - there should be a function that does that. now that i think about it all of the point cloud operators do this so i should clean that up.